flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorageForRecovery.java

```diff
@@ -0,0 +1,33 @@
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.api.common.operators.MailboxExecutor;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.changelog.ChangelogStateHandleStreamImpl;
+import org.apache.flink.runtime.state.changelog.StateChangelogHandleReader;
+import org.apache.flink.runtime.state.changelog.StateChangelogHandleStreamHandleReader;
+import org.apache.flink.runtime.state.changelog.StateChangelogStorage;
+import org.apache.flink.runtime.state.changelog.StateChangelogWriter;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Filesystem-based implementation of {@link StateChangelogStorage} just for recovery. */
+@Experimental
+@ThreadSafe
+public class FsStateChangelogStorageForRecovery
+        implements StateChangelogStorage<ChangelogStateHandleStreamImpl> {
+
+    @Override
+    public StateChangelogWriter<ChangelogStateHandleStreamImpl> createWriter(
+            String operatorID, KeyGroupRange keyGroupRange, MailboxExecutor mailboxExecutor) {
+        // The StateChangelogWriter will not be used when recovery
+        // under current design of Filesystem-based State Changelog Storage
+        throw new UnsupportedOperationException(
+                "createWriter is not supported for recovery from Changelog");
+    }
+
+    @Override
+    public StateChangelogHandleReader<ChangelogStateHandleStreamImpl> createReader() {
+        return new StateChangelogHandleStreamHandleReader(new StateChangeFormat());
+    }
+}
```

Comment:
The current PR hierarchy looks like this: `StateChangelogStorage -> FsStateChangelogStorageForRecovery -> FsStateChangelogStorageForRecovery`

How about splitting the top-level `StateChangelogStorage` into reader and writer and implementing them separately, so that:
1. We don't need to implement a method by throwing `UnsupportedOperationException`
2. We can return read-only interface from `StateChangelogStorageLoader.load`
3. It's easier to implement the needed part for tests

The new interface could be named as read-only, view, or something like that.
The existing interface can be completely separate write-only version, or it could extend the read-only counter-part. I think the latter is preferrable.

WDYT?

Reply:
Yeah, I agree. I just use StateChangelogStorageView to produce reader as you suggested.